docs/spec/auth/subkeys_proposal.md

```diff
@@ -0,0 +1,186 @@
+
+
+## Construction
+
+To begin, we define the struct that stores the metadata about a certain account.  This includes the pubkey itself, the routes it's permitted to sign transactions for, its daily allowance, and how much of its allowance has been used in the last 24 hour window.
+
+```golang
+type SubKeyMetadata struct {
+  PubKey               sdk.AccPubKey
+  PermissionedRoutes   []string
+  DailyFeeAllowance    sdk.Coins
+  DailyFeeUsed         sdk.Coins
+  Revoked              bool
+}
+```
+
+We create a new `Account` type called `SubKeyAccount` that stores a list of `SubKeyMetadata`s.
+
+```golang
+type SubKeyAccount struct {
+  Address       AccAddress
+  Coins         Coins
+  PubKey        sdk.AccPubKey
+  AccountNumber uint64
+  Sequence      uint64
+  SubKeys       []SubKeyMetadata
+}
+```
+
+In the `StdSignature` type we need to add a new field called `PubKeyIndex` which tells the anteHandler which SubKey to use to verify this signature.
+
+```golang
+// StdSignature represents a sig
+type StdSignature struct {
+  PubKey        crypto.PubKey
+  Signature     []byte
+  PubKeyIndex   uint
+}
+```
+
+If a `StdSignature.PubKeyIndex` is 0, this means that the account's "master" PubKey (the `SubKeyAccount.PubKey`) should be used to verify the signature.  If the `StdSignature.PubKeyIndex` is `> 0`, it verifies the signature using the corresponding **1-indexed** SubKey in the `SubKeyAccount.SubKeys` slice.  Note, the SubKeys slice is 1-indexed rather than 0-indexed because the 0 index is used to refer to the master PubKey.  See pseudocode below:
+
+```
+if stdSig.PubKeyIndex == 0 {
+  verify stdSig using acc.PubKey
+} else {
+  verify stdSig using acc.SubKeys[stdSig.PubKeyIndex - 1]
+}
+```
+
+In the AnteHandler, if a transaction comes from a SubKey, it verifies that the SubKey has not been revoked and it is permitted to use the msg routes it is trying to sign over.
+
+```
+if acc.SubKeys[stdSig.PubKeyIndex - 1].Revoked {
+  return ErrNotPermitted
+}
+if msg.Route not in subKeyMetadata.PermissionedRoutes {
+  return ErrNotPermitted
+}
+```
+
+When a transaction signed by a SubKey is accepted and has paid fees, we need to log the fees it has paid.  We increase the `SubKeyMetadata.DailyFeeUsed` field.
+
+If a new transaction from a subkey comes in and the `tx.Fee + subKeyMetadata.DailyFeeUsed > subKeyMetadata.DailyFeeAllowance`, then the transaction is rejected as the subkey is trying to exceed its daily fee allowance.
```

Comment:
what I mean here is that _if_ we support messages that only involve fees (without amounts), then we should add a check on the `PermissionedRoutes`.

Otherwise we should add both `DailyAmountAllowance` and `DailyAmountUsed` (or merge it with the `DailyFeeAllowance`/`DailyFeeUsed`)

Reply:
Yeah, I think a DailyAmountAllowance is really good idea, so that you can limit SubKeys to only spending a certain amount.  However, I think is a more complex thing, because you "spends" can happen in many different ways across different modules (bank sends, gov deposit, etc).

Nice thing about checking just DailyFeeAllowance is it is only checked in one spot in the ante handler, and that's it, doesn't affect any other module.

I think DailyFeeAllowances is a good starting spot.  We can do the more general AccountAllowance in the future; I think it needs a more complex feature called "SubAccounts".